lib/lvelements/src/parseddocument.cpp

```diff
@@ -6,19 +6,464 @@
+            {
+                auto start = ts_node_start_byte(ts_node_child(curr, ip));
+                auto end = ts_node_end_byte(ts_node_child(curr, ip));
+                if (start < position)
+                {
+                    if (end < position)
+                    {
+                        expressionPath.push_back(SourceRange(start, end - start));
+                    }
+                    else {
+                        expressionPath.push_back(SourceRange(start, position-start));
+                        break;
+                    }
+                }
+            }
+
+            return CursorContext(context,expressionPath);
+        }
+
+        if (strcmp(type, "string") == 0)
+        {
+            context = CursorContext::InStringLiteral;
+        }
+
+    }
+
+
+    return CursorContext(context, expressionPath);
+}
+
+void ParsedDocument::treePath(LanguageParser::AST *ast, int position, std::vector<TSNode> &result)
+{
+    TSTree* tree = reinterpret_cast<TSTree*>(ast);
+    TSNode root_node = ts_tree_root_node(tree);
+
+    TSNode current = root_node;
+    bool found = true;
+    while (found)
+    {
+        auto count = ts_node_child_count(current);
+        found = false;
+        for (int idx = 0; idx < count; ++idx)
+        {
+            TSNode child = ts_node_child(current, idx);
+            auto start = ts_node_start_byte(child);
+            auto end = ts_node_end_byte(child);
+
+            if (start <= position && position <= end)
+            {
+                result.push_back(child);
+                current = child;
+                found = true;
+                break;
+            }
+        }
+    }
+}
+
+TypeInfo::Ptr ParsedDocument::extractType(const std::string& source, TSNode node)
+{
+    Utf8                      typeName;
+    // Utf8                      className; = ?
+    Utf8                      inherits = "Element";
+    // std::vector<EnumInfo>     enums; = none
+    std::vector<PropertyInfo> properties;
+    std::vector<FunctionInfo> functions;
+    std::vector<FunctionInfo> events;
+    // bool                      isCreatable; = false
+    bool                      isInstance = false;
+
+    if (strcmp(ts_node_type(node), "new_component_expression") == 0)
+    {
+        if (strcmp(ts_node_type(ts_node_child(node, 0)), "component_instance") != 0)
+            return nullptr;
+        isInstance = true;
+    }
+
+
+
+    auto count = ts_node_child_count(node);
+    for (int i = 0; i < count; i++)
+    {
+        TSNode cdChild = ts_node_child(node, i);
+
+        if (strcmp(ts_node_type(cdChild), "identifier") == 0)
+        {
+            auto start = ts_node_start_byte(cdChild);
+            auto length = ts_node_end_byte(cdChild) - start;
+            typeName = source.substr(start, length);
+
+
+        }
+        else if (strcmp(ts_node_type(cdChild), "component_heritage") == 0)
+        {
+
+            auto start = ts_node_start_byte(ts_node_child(cdChild, 1));
```

Comment:
I noticed this patter a lot here:

```cpp
start = ...(node)
length = ...(node)
content = source.substr(...)
```

You could have a slice method for this:

```
std::string slice(const string source, const TSNode& node)
```